### PR TITLE
perf: using Node.js compile cache

### PR DESCRIPTION
**Pull request by aladdin-add**
it was introduced in node.js v22.8.0
https://nodejs.org/en/blog/release/v22.8.0#new-js-api-for-compile-cache

fixes https://github.com/eslint/eslint/issues/18968

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ x] Other, please explain:

#### What changes did you make? (Give an overview)

tested on
* ESLint: v9.12.0
* Node version: v22.9.0
* Operating System: darwin 24.1.0

before the change: 
```bash
Loading:
  Load performance Run #1:  181.045881ms
  Load performance Run #2:  188.315573ms
  Load performance Run #3:  411.307467ms
  Load performance Run #4:  208.031565ms
  Load performance Run #5:  240.104659ms

  Load Performance median:  208.031565ms
```
after the change(2rd time): 
```bash
Loading:
  Load performance Run #1:  123.129709ms
  Load performance Run #2:  136.558748ms
  Load performance Run #3:  134.139566ms
  Load performance Run #4:  136.995985ms
  Load performance Run #5:  206.45432ms

  Load Performance median:  136.558748ms
```

#### Is there anything you'd like reviewers to focus on?
`module.enableCompileCache()` is still experimental [Stability: 1](https://nodejs.org/docs/latest/api/documentation.html#stability-index).1 - Active Development

https://nodejs.org/docs/latest/api/module.html#moduleenablecompilecachecachedir

### CODE REVIEW

File: bin/eslint.js [hunk 1]
```diff
@@ -9,6 +9,11 @@
 
 "use strict";
 
+const mod = require("node:module");
+
+// to use V8's code cache to speed up instantiation time
+mod.enableCompileCache?.();
+
```
Comment (snitin315) [inline bin/eslint.js:16]: Would it work for developers using the Node.js API and not CLI?

Comment (aladdin-add) [reply bin/eslint.js:16]: it's intentional - see https://github.com/eslint/eslint/pull/11921#pullrequestreview-256057049

Comment (snitin315) [reply bin/eslint.js:16]: Makes sense 👍🏻 .

**Review by snitin315 — APPROVED** [review]

**Comment by snitin315** [comment]
LGTM. Awaiting 2nd review

**Comment by mdjermanovic** [comment]
> `module.enableCompileCache()` is still experimental [Stability: 1](https://nodejs.org/docs/latest/api/documentation.html#stability-index).1 - Active Development

Then maybe we shouldn't use it yet? Description for Stability: 1 says "Use of the feature is not recommended in production environments".

**Comment by snitin315** [comment]
> > `module.enableCompileCache()` is still experimental [Stability: 1](https://nodejs.org/docs/latest/api/documentation.html#stability-index).1 - Active Development
> 
> Then maybe we shouldn't use it yet? Description for Stability: 1 says "Use of the feature is not recommended in production environments".

I think it should be fine to use, especially since [TypeScript recently implemented](https://github.com/microsoft/TypeScript/pull/59720) this too. We can always keep an eye out for any potential changes or improvements.

**Comment by mdjermanovic** [comment]
> I think it should be fine to use, especially since [TypeScript recently implemented](https://github.com/microsoft/TypeScript/pull/59720) this too. We can always keep an eye out for any potential changes or improvements.

Looks like they wrapped the `enableCompileCache()` call in a try-catch?

https://github.com/microsoft/TypeScript/blob/40caf3431986884b3da7a607de22ced5d364d9fc/Herebyfile.mjs#L332-L337

And there's also a config option to disable it:

https://github.com/microsoft/TypeScript/blob/40caf3431986884b3da7a607de22ced5d364d9fc/Herebyfile.mjs#L320

File: eslint.config.js [hunk 2]
```diff
@@ -305,5 +305,17 @@ module.exports = [
...
+    {
+        name: "eslint/bin",
+        files: ["bin/eslint.js"],
+        rules: {
+
+            /*
+             * it was introduced in Node.js v22.8.0
+             * refs: https://nodejs.org/en/blog/release/v22.8.0#new-js-api-for-compile-cache
+             */
+            "n/no-unsupported-features/node-builtins": [2, { ignores: ["module.enableCompileCache"] }]
```
Comment (mdjermanovic) [inline eslint.config.js:318]: Can you also bump `eslint-plugin-n` to the latest version to ensure this option is available in `ignores`?

**Comment by snitin315** [comment]
> > I think it should be fine to use, especially since [TypeScript recently implemented](https://github.com/microsoft/TypeScript/pull/59720) this too. We can always keep an eye out for any potential changes or improvements.
> 
> Looks like they wrapped the `enableCompileCache()` call in a try-catch?
> 
> https://github.com/microsoft/TypeScript/blob/40caf3431986884b3da7a607de22ced5d364d9fc/Herebyfile.mjs#L332-L337
> 
> And there's also a config option to disable it:
> 
> https://github.com/microsoft/TypeScript/blob/40caf3431986884b3da7a607de22ced5d364d9fc/Herebyfile.mjs#L320

Hmm, can we also implement this with a flag, something like `eslint --flag enable-compile-cache`?

**Comment by aladdin-add** [comment]
It depends on what the result of a failed call is - if it's that in some cases the cache won't take effect without throwing an error - I think that's perfectly acceptable!

**Comment by mdjermanovic** [comment]
> It depends on what the result of a failed call is - if it's that in some cases the cache won't take effect, and _Not_ throwing an error - I think that's perfectly acceptable!

If the call ends up having no effect, I also think that's perfectly acceptable. I'm only concerned if module loading will start throwing errors or malfunctioning in any other way at some later point, as this is an experimental feature.

**Comment by aladdin-add** [comment]
its docs said:

> Since compile cache is supposed to be a quiet optimization that is not required for the application to be functional, this method is designed to not throw any exception when the compile cache cannot be enabled.

https://nodejs.org/docs/latest/api/module.html#moduleenablecompilecachecachedir

Just in case having an issue,  it's no harm to put it in a `try...catch`. wdyt?

**Comment by mdjermanovic** [comment]
> its docs said:
> 
> > Since compile cache is supposed to be a quiet optimization that is not required for the application to be functional, this method is designed to not throw any exception when the compile cache cannot be enabled.
> 
> https://nodejs.org/docs/latest/api/module.html#moduleenablecompilecachecachedir

I think the statement refers to the API: `enableCompileCache()` call will not throw an error when the compile cache cannot be enabled. But since it's still marked as an experimental feature, I'm not sure if it's guaranteed that everything will work properly when the compile cache is enabled.

**Comment by mdjermanovic** [comment]
> Hmm, can we also implement this with a flag, something like `eslint --flag enable-compile-cache`?

That might be a good option, though I think users can already manually enable this by setting the `NODE_COMPILE_CACHE` environment variable.

**Comment by aladdin-add** [comment]
well, now we have 3 choices：

1. enable it by default, and keep an eye out for any potential changes or improvements..
2. wait a while for it to stabilize before enabling it.
3. enable it when specifying ` --flag enable-compile-cache` (not enabled by default)

my preference is 2 > 1 > 3.

**Comment by nzakas** [comment]
The API is the part that's experimental, the feature has been around for a while but only enabled via environment variable. So the compile cache itself isn't experimental.

**Comment by mdjermanovic** [comment]
> The API is the part that's experimental, the feature has been around for a while but only enabled via environment variable. So the compile cache itself isn't experimental.

In that case, it seems safe to merge.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**